apollo-router/src/services/layers/query_analysis.rs

```diff
@@ -0,0 +1,150 @@
+use std::sync::Arc;
+
+use apollo_compiler::ApolloCompiler;
+use apollo_compiler::FileId;
+use apollo_compiler::HirDatabase;
+use http::StatusCode;
+use lru::LruCache;
+use tokio::sync::Mutex;
+
+use crate::services::SupergraphRequest;
+use crate::services::SupergraphResponse;
+use crate::spec::query::Operation;
+use crate::spec::Query;
+use crate::spec::Schema;
+use crate::Configuration;
+
+/// [`Layer`] for QueryParsing implementation.
+#[derive(Clone)]
+pub(crate) struct QueryAnalysisLayer {
+    /// set to None if QueryParsing is disabled
+    schema: Arc<Schema>,
+    configuration: Arc<Configuration>,
+    cache: Arc<Mutex<LruCache<QueryAnalysisKey, Arc<Mutex<ApolloCompiler>>>>>,
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+struct QueryAnalysisKey {
+    query: String,
+    operation_name: Option<String>,
+}
+
+impl QueryAnalysisLayer {
+    pub(crate) async fn new(schema: Arc<Schema>, configuration: Arc<Configuration>) -> Self {
+        let mut cache = configuration
+            .supergraph
+            .query_planning
+            .experimental_cache
+            .clone();
+        cache.redis = None;
```

Comment:
Nit: why these two lines if only `in_memory.limit` ends up used?

Reply:
because before, I was trying to use `DeduplicatingCache` which took the whole cache config structure, but I did not want it to set up redis for this. fixed in [d579c33](https://github.com/apollographql/router/pull/3200/commits/d579c334219dc7d9b8c46c9e39935f0e08dea02a)